src/Marten/Events/Daemon/ProjectionDaemon.cs

```diff
@@ -1,53 +1,52 @@
-#nullable enable
 using System;
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using JasperFx.Core;
-using JasperFx.Core.Reflection;
 using Marten.Events.Daemon.HighWater;
-using Marten.Events.Daemon.Progress;
+using Marten.Events.Daemon.Internals;
 using Marten.Events.Daemon.Resiliency;
-using Marten.Events.Projections;
-using Marten.Exceptions;
 using Marten.Services;
 using Marten.Storage;
 using Microsoft.Extensions.Logging;
+using Weasel.Core;
 
 namespace Marten.Events.Daemon;
 
-/// <summary>
-///     The main class for running asynchronous projections
-/// </summary>
-internal class ProjectionDaemon: IProjectionDaemon
+public partial class ProjectionDaemon : IProjectionDaemon, IObserver<ShardState>, IDaemonRuntime
 {
-    private readonly Dictionary<string, ShardAgent> _agents = new();
-    private readonly HighWaterAgent _highWater;
-    private readonly ILogger _logger;
     private readonly DocumentStore _store;
-    private CancellationTokenSource _cancellation;
-    private INodeCoordinator? _coordinator;
-
-    private bool _isStopping;
-    private readonly RetryBlock<DeadLetterEvent> _deadLetterBlock;
+    private readonly IAgentFactory _factory;
+    private readonly List<ISubscriptionAgent> _active = new();
```

Comment:
**CONCERN:** Shouldn't we use here `ImHashMap`? That'd potentially limit the scenarios with running multithreaded access when starting/stopping/detecting the shard agents in the methods below (Start, Stop, etc.). This can be important in terms of pausing and restarting active daemon shards.

Reply:
Good catch. See https://github.com/JasperFx/marten/issues/2967 for follow up work. Yeah, I think you're right about the potential for concurrent access. The `ImHashMap` might be appropriate (or ImList instead), but I'd still lock *write* access to it regardless. In this new model, we're scrapping and rebuilding the `SubscriptionAgent` completely instead of just pausing/restarting. Made for a little bit simpler model